[JSC] Add global replace caching

Constellation opened the pull request:
#### 0e3fcb9531e248c468366393fd4f2127ea5dcb56
<pre>
[JSC] Add global replace caching
<a href="https://bugs.webkit.org/show_bug.cgi?id=257826">https://bugs.webkit.org/show_bug.cgi?id=257826</a>
rdar://110417641

Reviewed by Michael Saboff.

This patch adds String.replace(/regexp-global/g, func) result cache, which is in the same fashion to V8&apos;s cache.
We cache the result only when the subject string length is &gt;= 0x1000, aligned to V8&apos;s behavior.
This cache is cleared on every GC, so it does not affect on memory usage in practice: once GC happens, everything gets cleared,
so it is not retaining things.

* JSTests/stress/v8-regexp-results-cache.js: Added.
* JSTests/stress/v8-string-replace-cache.js: Added.
(replaceTest):
(r):
(i.r):
(string_appeared_here.string_appeared_here.replace_obj.toString):
(search_obj.toString):
(replace_obj_side_effects.toString):
(testIndices99):
(testIndices59):
(const.fake_replacer.Symbol.toPrimitive):
* JSTests/stress/v8-string-replace.js: Added.
(replaceTest):
(r):
(i.r):
(string_appeared_here.string_appeared_here.replace_obj.toString):
(search_obj.toString):
(replace_obj_side_effects.toString):
(testIndices99):
(testIndices59):
(const.fake_replacer.Symbol.toPrimitive):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::finalize):
* Source/JavaScriptCore/runtime/ArgList.h:
* Source/JavaScriptCore/runtime/JSImmutableButterfly.cpp:
(JSC::JSImmutableButterfly::tryCreateFromArgList):
* Source/JavaScriptCore/runtime/JSImmutableButterfly.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/RegExpGlobalData.h:
(JSC::RegExpGlobalData::ovector const):
* Source/JavaScriptCore/runtime/RegExpGlobalDataInlines.h:
(JSC::RegExpGlobalData::resetResultFromCache):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::replaceUsingRegExpSearchWithCache):
(JSC::replaceUsingRegExpSearch):
* Source/JavaScriptCore/runtime/StringReplaceCache.h: Copied from Source/JavaScriptCore/runtime/RegExpGlobalData.h.
(JSC::StringReplaceCache::clear):
* Source/JavaScriptCore/runtime/StringReplaceCacheInlines.h: Added.
(JSC::StringReplaceCache::get):
(JSC::StringReplaceCache::set):
* Source/JavaScriptCore/runtime/VM.h:

Canonical link: <a href="https://commits.webkit.org/264967@main">https://commits.webkit.org/264967@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a9b5b56f5f1404eae0c2fa1c10e293a9bd29a799

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9286 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9567 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9795 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10947 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9175 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11552 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9530 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12059 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9435 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10366 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8019 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11106 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7662 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8477 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15929 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/7966 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8760 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8628 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11968 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/8893 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9120 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7431 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/9462 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8336 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/2281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12560 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/9702 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1067 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8881 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2382 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

msaboff commented on Source/JavaScriptCore/runtime/StringReplaceCacheInlines.h:
```diff
@@ -0,0 +1,89 @@
+                entry2.m_result = result;
+            } else {
+                entry2 = { };
+                entry1.m_subject = subjectImpl;
+                entry1.m_regExp = regExp;
+                entry1.m_lastMatch = lastMatch;
+                entry1.m_result = result;
```
Is this a bug given you clear entry2, but modify entry1?  Or is this just your replacement policy?

Constellation replied:
Yeah, it is intentional. The behavior is clear both and use entry1.

msaboff replied:
Why wouldn't you want to keep entry2?

Constellation replied:
Either is fine, but for now, aligning to the V8's behavior.

msaboff reviewed (approved):
r=me with question.

Constellation committed:
[JSC] Add global replace caching
https://bugs.webkit.org/show_bug.cgi?id=257826
rdar://110417641

Reviewed by Michael Saboff.

This patch adds String.replace(/regexp-global/g, func) result cache, which is in the same fashion to V8's cache.
We cache the result only when the subject string length is >= 0x1000, aligned to V8's behavior.
This cache is cleared on every GC, so it does not affect on memory usage in practice: once GC happens, everything gets cleared,
so it is not retaining things.

* JSTests/stress/v8-regexp-results-cache.js: Added.
* JSTests/stress/v8-string-replace-cache.js: Added.
(replaceTest):
(r):
(i.r):
(string_appeared_here.string_appeared_here.replace_obj.toString):
(search_obj.toString):
(replace_obj_side_effects.toString):
(testIndices99):
(testIndices59):
(const.fake_replacer.Symbol.toPrimitive):
* JSTests/stress/v8-string-replace.js: Added.
(replaceTest):
(r):
(i.r):
(string_appeared_here.string_appeared_here.replace_obj.toString):
(search_obj.toString):
(replace_obj_side_effects.toString):
(testIndices99):
(testIndices59):
(const.fake_replacer.Symbol.toPrimitive):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::finalize):
* Source/JavaScriptCore/runtime/ArgList.h:
* Source/JavaScriptCore/runtime/JSImmutableButterfly.cpp:
(JSC::JSImmutableButterfly::tryCreateFromArgList):
* Source/JavaScriptCore/runtime/JSImmutableButterfly.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/RegExpGlobalData.h:
(JSC::RegExpGlobalData::ovector const):
* Source/JavaScriptCore/runtime/RegExpGlobalDataInlines.h:
(JSC::RegExpGlobalData::resetResultFromCache):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::replaceUsingRegExpSearchWithCache):
(JSC::replaceUsingRegExpSearch):
* Source/JavaScriptCore/runtime/StringReplaceCache.h: Copied from Source/JavaScriptCore/runtime/RegExpGlobalData.h.
(JSC::StringReplaceCache::clear):
* Source/JavaScriptCore/runtime/StringReplaceCacheInlines.h: Added.
(JSC::StringReplaceCache::get):
(JSC::StringReplaceCache::set):
* Source/JavaScriptCore/runtime/VM.h:

Canonical link: https://commits.webkit.org/264967@main
